# eda에 앞서 고민
1. 고유명사: 토큰화가 정상적으로 되지 않거나 형태소 분석에서 큰 오류가 생길 위험이 높아 어려울 것이다.
    - 고유명사는 따옴표로 감싼다.
    - 
2. 동의어, 유의어, 반의어 + 부정 표현: 정확히 일치하는 단어에 비해 크게 어려울 것이다. 증강 아이디어로 활용해도 좋을 듯하다.
3. 추론 문제는 어떻게?
4. (증강시) context에 정답 단어가 여러번 나올 때는 직접 어디 맥락인지 지정해주어야 한다.

In [2]:
import os
import json
import pandas as pd
from datasets import load_from_disk

/opt/miniconda3/envs/MRC/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [3]:
current_dir = os.getcwd()
parent_dir = os.path.dirname(current_dir)

# Arrow 파일 경로 설정
dataset = load_from_disk(os.path.join(parent_dir, "data", "train_dataset"))
test_dataset = load_from_disk(os.path.join(parent_dir, "data", "test_dataset"))

train_df = pd.DataFrame(dataset["train"])
valid_df = pd.DataFrame(dataset["validation"])
test_df = pd.DataFrame(test_dataset["validation"])

print(f"Train length : {len(train_df)}")
print(f"Valid length : {len(valid_df)}")
print(f"Test length : {len(test_df)}")

Train length : 3952
Valid length : 240
Test length : 600


In [4]:
current_dir = os.getcwd()
parent_dir = os.path.dirname(current_dir)
file_path = os.path.join(parent_dir, 'data', "wikipedia_documents.json")

with open(file_path, "r", encoding="utf-8") as json_file:
    json_data = json.load(json_file)
    
df_wiki=pd.read_json(file_path).transpose()

In [5]:
df_wiki.head()

text corpus_source   url  \
0  이 문서는 나라 목록이며, 전 세계 206개 나라의 각 현황과 주권 승인 정보를 개...         위키피디아  TODO   
1  이 목록에 실린 국가 기준은 1933년 몬테비데오 협약 1장을 참고로 하였다. 협정...         위키피디아  TODO   
2  현 서울특별시 종로구 서린동 (구 일제 강점기 경기도 경성부 서린정) 출신이다. 친...         위키피디아  TODO   
3  아오조라 문고(靑空文庫, あおぞらぶんこ|아오조라 분고)는 ‘일본어판 구텐베르크 프로...         위키피디아  TODO   
4  저자 사망 이후 50년이 지나 저작권이 소멸한 메이지 시대부터 쇼와 시대 초기까지의...         위키피디아  TODO   

  domain    title author  html document_id  
0   None    나라 목록   None  None           0  
1   None    나라 목록   None  None           1  
2   None      백남준   None  None           2  
3   None  아오조라 문고   None  None           3  
4   None  아오조라 문고   None  None           4

In [6]:
train_df.head()

title                                            context  \
0    미국 상원  미국 상의원 또는 미국 상원(United States Senate)은 양원제인 미국...   
1   인사조직관리  '근대적 경영학' 또는 '고전적 경영학'에서 현대적 경영학으로 전환되는 시기는 19...   
2      강희제  강희제는 강화된 황권으로 거의 황제 중심의 독단적으로 나라를 이끌어 갔기에 자칫 전...   
3   금동삼존불감  불상을 모시기 위해 나무나 돌, 쇠 등을 깎아 일반적인 건축물보다 작은 규모로 만든...   
4  계사명 사리구  동아대학교박물관에서 소장하고 있는 계사명 사리구는 총 4개의 용기로 구성된 조선후기...   

                                  question            id  \
0         대통령을 포함한 미국의 행정부 견제권을 갖는 국가 기관은?  mrc-1-000067   
1                   현대적 인사조직관리의 시발점이 된 책은?  mrc-0-004397   
2           강희제가 1717년에 쓴 글은 누구를 위해 쓰여졌는가?  mrc-1-000362   
3  11~12세기에 제작된 본존불은 보통 어떤 나라의 특징이 전파되었나요?  mrc-0-001510   
4               명문이 적힌 유물을 구성하는 그릇의 총 개수는?  mrc-0-000823   

                                         answers  document_id  \
0        {'answer_start': [235], 'text': ['하원']}        18293   
1  {'answer_start': [212], 'text': ['《경영의 실제》']}        51638   
2        {'answer_start': [510], 'text': ['백성']}         5028   
3        {'answer_start': [625], 'text': ['중국']}        34146   
4         {'answer_start': [30], 'text': ['4개']}        47334   

   __index_level_0__  
0                 42  
1               2873  
2                230  
3                992  
4                548

In [7]:
valid_df.head()

title                                            context  \
0    전효숙  순천여자고등학교 졸업, 1973년 이화여자대학교를 졸업하고 1975년 제17회 사법...   
1  스토우빌선  요크 카운티 동쪽에 처음으로 여객 열차 운행이 시작한 시점은 1868년 토론토 & ...   
2     촐라  촐라 정부\n 촐라의 정부 체제는 전제군주제였으며,2001 촐라의 군주는 절대적인 ...   
3  백조 기사  프랑스의 십자군 무훈시는 1099년 예루살렘 왕국의 통치자가 된 고드프루아 드 부용...   
4  탈관료제화  탈관료제화는 현대사회에서 관료제 성격이 약화되는 현상이다. 현대사회에서 관료제는 약...   

                               question            id  \
0        처음으로 부실 경영인에 대한 보상 선고를 받은 회사는?  mrc-0-003264   
1  스카버러 남쪽과 코보콘그 마을의 철도 노선이 처음 연장된 연도는?  mrc-0-004762   
2       촌락에서 운영 위원 후보자 이름을 쓰기위해 사용된 것은?  mrc-1-001810   
3              로타이르가 백조를 구하기 위해 사용한 것은?  mrc-1-000219   
4    의견을 자유롭게 나누는 것은 조직 내 어떤 관계에서 가능한가?  mrc-1-000285   

                                       answers  document_id  __index_level_0__  
0    {'answer_start': [284], 'text': ['한보철강']}         9027               2146  
1   {'answer_start': [146], 'text': ['1871년']}        51765               3106  
2     {'answer_start': [517], 'text': ['나뭇잎']}        15694               1180  
3    {'answer_start': [1109], 'text': ['금대야']}        59536                141  
4  {'answer_start': [386], 'text': ['수평적 관계']}        32991                184

## ID나 __index_level_0__는 무슨 의미일까?

In [7]:
train_id_prefix = train_df["id"].map(lambda x: x[:5])
train_id_prefix.value_counts()

id
mrc-0    2957
mrc-1     995
Name: count, dtype: int64

In [8]:
valid_id_prefix  = valid_df["id"].map(lambda x: x[:5])
valid_id_prefix.value_counts()

id
mrc-0    176
mrc-1     64
Name: count, dtype: int64

__index_level_0__은 CSV 파일 등을 pyarrow로 읽어올 때 index를 보존하는 경우 생성된다.

In [9]:
train_df = train_df.drop(["id", "__index_level_0__"],axis=1)

In [10]:
valid_df = valid_df.drop(["id", "__index_level_0__"], axis=1)

## Wikipedia에 존재하는 데이터인가?

In [11]:
def is_df_in_wiki(df):
    no_wiki = [
        row
        for _, row in df.iterrows()
        if row["document_id"] not in df_wiki["document_id"]
    ]
    return pd.DataFrame(no_wiki, columns=df.columns)

In [12]:
print(
    len(is_df_in_wiki(train_df)), 
    len(is_df_in_wiki(valid_df))
)

0 0


## answer가 context에 존재하는가?

In [13]:
def is_answer_in_context(df):
    answer_in_question = [
        row for _, row in df.iterrows() if row["answers"]["text"][0] not in row["context"]
    ]
    return pd.DataFrame(answer_in_question, columns=df.columns)

In [14]:
is_answer_in_context(train_df)

Empty DataFrame
Columns: [title, context, question, answers, document_id]
Index: []

In [15]:
is_answer_in_context(valid_df)

Empty DataFrame
Columns: [title, context, question, answers, document_id]
Index: []

## answer_start가 정확한가?

In [85]:
# 정답 검증 함수
def verify_answer(row):
    context = row["context"]
    answer_text = row["answers"]["text"][0]
    answer_start = row["answers"]["answer_start"][0]

    # answer_start부터 answer_text 길이만큼의 문장을 추출
    extracted_answer = context[answer_start : answer_start + len(answer_text)]

    # 추출한 정답과 answer_text가 동일한지 검증
    return extracted_answer == answer_text


# 잘못된 정답을 포함한 행만 모아서 새로운 DataFrame 반환
def get_incorrect_rows(df):
    # 각 행에 대해 검증 수행하여 False인 행만 필터링
    incorrect_df = df[~df.apply(verify_answer, axis=1)]
    return incorrect_df

In [11]:
# find 함수 사용하여 검증할 수도 있음
context = "오늘은 2024년 10월 16일 입니다."
answer_text = "16일"
answer_start = context.find(answer_text)
print(context[answer_start : answer_start + len(answer_text)], answer_start)

16일 14


In [86]:
train_incorrect = get_incorrect_rows(train_df)
valid_incorrect = get_incorrect_rows(valid_df)

In [87]:
train_incorrect

Empty DataFrame
Columns: [title, context, question, answers, document_id, answer_sentence]
Index: []

In [88]:
valid_incorrect

Empty DataFrame
Columns: [title, context, question, answers, document_id, answer_sentence]
Index: []

## answer가 question에도 존재하는가?

In [16]:
def is_answer_in_question(df):
    answer_in_question = [
        row for _, row in df.iterrows() if row["answers"]["text"][0] in row["question"]
    ]
    return pd.DataFrame(answer_in_question, columns=df.columns)

In [17]:
is_answer_in_question(train_df)

title  \
147                     교황 클레멘스 2세   
275                         아베 오사무   
404                      고창 오거리 당산   
614         마이어로위츠 이야기 (제대로 고른 신작)   
631                     대한민국의 보수주의   
927                     탈염과 완충액 교환   
945   네덜란드 대 대한민국 (1998년 FIFA 월드컵)   
1451                    창녕 관룡사 석장승   
1584                         규소 수지   
1634              피아노 협주곡 1번 (베토벤)   
1910                      캐나디안 케이퍼   
2687                 사망유희 (인터넷 토론)   
3460                  칠보사 목조관음보살좌상   
3523         시민적 및 정치적 권리에 관한 국제규약   
3727                          토하라인   
3922                         수소 결합   

                                                context  \
147   오늘날 독일 니더작센주 호른부르크 태생인 그는 몰스레벤 및 호른부르크 백작 콘라드와...   
275   1980년 프로 야구 드래프트 회의에서 세이부 라이온스로부터 4순위 지명을 받아 입...   
404   고창읍 중앙동에 위치한 당산이다. 당산은 민간신앙에서 신이 있다고 믿고 섬기는 것으...   
614   마이어로위츠 가족의 맏아들 대니는 최근 아내와 이혼한 후 뉴욕 맨해튼에 있는 아버지...   
631   보수주의\n*진정한 자유민주주의를 추구하는 게 아니라는 비판\n** 자유민주주의는 ...   
927   작은 (4mL 미만의) 부피에 대해 겔 여과를 수행하기위한 여러가지 일반적인 방식이...   
945   1998년 대회에서 대한민국은 네덜란드, 멕시코, 벨기에와 함께 E조에 편성되었다....   
1451  관룡사로 가는 오솔길 양쪽에 서 있는 한 쌍의 돌장승이다.\n\n사찰 입구에 두는 ...   
1584  규소 수지(硅素樹脂)는 실리콘 물질의 일종이다.\n\n이산화규소는 1개의 규소 원자...   
1634  하이든 밑에서 본격적으로 작곡을 배우기 위해 1792년 본에서 빈으로 자리를 옮긴 ...   
1910  1979년, 격해지는 반미 시위 가운데 이란 시위대가 주 이란 미국 대사관 벽을 넘...   
2687  2012년 10월 19일 진중권은 시사토론에 출연해 NLL과 정수장학회에 대한 토론...   
3460  칠보사 목조관음보살좌상은 약간 구부린 자세에 넓은 어깨와 무릅이 높고 넓어 안정감이...   
3523  세계 인권 선언은 국제관습법이 되어서 모든 나라에 법률적 효력이 있다는 일각의 주장...   
3727  기원전 2세기 흉노로 인해 타림분지에서 인도로 이동하였다. 스키타이족을 비롯한 이란...   
3922  물에서의 수소 결합은 H-O-H···O-2H 형태로 나타난다.\n\n물 분자에서 2...   

                                             question  \
147                       아물라드의 아들이 밤베르크의 주교로 서임된 해는?   
275           9월 23일 세이부 돔에서 진행된 경기에서 긴테쓰의 상대팀은 어디였나?   
404                    중앙동·중거리·하거리 할아버지당 중 가장 높은 당산은?   
614                         폴이 어릴 적 성적으로 흥분을 느꼈던 대상은?   
631                     자유민주주의에서 사람들 모두가 보장받아야 하는 것은?   
927                크로마토 그래피 카트리지가 한 번에 사용 가능한 샘플의 수는?   
945                         한국 대 멕시코 경기에서 선제골을 넣은 쪽은?   
1451                      여장승과 남장승 중에 키가 더 큰 것은 무엇인가?   
1584                       실리콘을 실리콘유로 만들기 위해 거치는 과정은?   
1634              두 번째, 세 번째 피아노 협주곡 중 먼저 세상에 알려진 곡은?   
1910                 앤더스 그룹은 시위하는 집단을 피해 누구의 집으로 갔는가?   
2687                사람들은 간결과 진중권의 토론에서 누가 이겼다고 생각하는가?   
3460  창녕 포교당 목조여래좌상과 대구 동화사 대웅전 목조삼세불좌상 중 먼저 만들어진 것은?   
3523     세계인권선언과 국제인권규약을 비교했을 때 어떤 것이 더 세세한 규약을 담고있나?   
3727                 토하라어는 인도이란어와 켈트어 중에서 무엇과 더 유사한가?   
3922              밤이 되면 공기가 위로 올라가는 곳은 육지와 바다 중 어디인가?   

                                                answers  document_id  
147            {'answer_start': [53], 'text': ['아물라드']}         6736  
275            {'answer_start': [681], 'text': ['세이부']}        36873  
404      {'answer_start': [415], 'text': ['하거리 할아버지당']}        45428  
614              {'answer_start': [918], 'text': ['폴']}        47869  
631              {'answer_start': [51], 'text': ['자유']}        29784  
927           {'answer_start': [641], 'text': ['한 샘플']}        53185  
945            {'answer_start': [1317], 'text': ['한국']}        51777  
1451           {'answer_start': [336], 'text': ['여장승']}        39893  
1584           {'answer_start': [190], 'text': ['실리콘']}        56358  
1634  {'answer_start': [255], 'text': ['세 번째 피아노 협주곡']}        58986  
1910           {'answer_start': [314], 'text': ['앤더스']}        33784  
2687           {'answer_start': [546], 'text': ['진중권']}        35704  
3460  {'answer_start': [447], 'text': ['대구 동화사 대웅전 목...        47336  
3523        {'answer_start': [165], 'text': ['국제인권규약']}        28063  
3727           {'answer_start': [120], 'text': ['켈트어']}        14242  
3922            {'answer_start': [627], 'text': ['바다']}        10486

In [18]:
is_answer_in_question(valid_df)

title                                            context  \
35  착생식물  해양 착생식물의 생태는 지상 착생식물의 생태와는 다르다. 해양 생태계의 착생식물은 ...   

                               question  \
35  해초나 조류 표면에서 자라는 유기체 중 가장 비율이 높은 것은?   

                                    answers  document_id  
35  {'answer_start': [180], 'text': ['조류']}        51188

# question이 객관식 형태인가?

In [19]:
def is_multiple_choice_question(df):
    answer_in_question = [
        row
        for _, row in df.iterrows()
        if " 중 " in row["question"] or " 중에" in row["question"]
    ]
    return pd.DataFrame(answer_in_question, columns=df.columns)

In [20]:
is_multiple_choice_question(train_df)

title                                            context  \
84           유달산  유달산(儒達山)은 전라남도 목포시 유달동, 목원동, 죽교동에 있는 높이 228미터의...   
106          균혈증  혈액 감염에 대한 면역반응|패혈증\n \n균혈증(菌血症, bacteremia, ba...   
144     원정 (군주제)  에도 시대에 들어 금중병공가제법도(禁中竝公家諸法度)를 제정한 에도 막부의 조정 개입...   
180          허계임  그녀는 경기도 용인의 한 이교도 집안에서 태어났다. 그녀는 경기도 봉천리에 있는 한...   
183          천연두  출혈성 천연두는 피부, 점막, 소화관에 광범한 내출혈을 동반하는 심각한 천연두 유형...   
...          ...                                                ...   
3745      삼국지 IX  다시 군주제로 변환했고, 10일 단위의 반(half) 리얼타임제를 도입했다. 도시 ...   
3746  충주 조동리 고인돌  지석묘는 청동기시대의 대표적인 무덤으로 고인돌이라고도 부른다. 우리나라의 고인돌은 ...   
3748       개방경지제  중세 시기 대상재배의 전형적인 삼경법은 밭을 셋으로 나눠 첫번째 밭에는 봄에 보리,...   
3917        차바하르  파키스탄은 중국과 손잡고 차바하르항에서 동쪽으로 불과 약 70 km에 떨어진 과다르...   
3922       수소 결합  물에서의 수소 결합은 H-O-H···O-2H 형태로 나타난다.\n\n물 분자에서 2...   

                                               question  \
84                           가수 이난영의 노래 중 기념비까지 세워진 것은?   
106                         치과 시술 중 혈관 속으로 들어올 수 있는 것은?   
144                      일본의 원정 중 가장 최근의 원정은 언제 실행되었는가?   
180                   허계임과 그녀의 두 딸 중 가장 먼저 순교한 것은 누구인가?   
183            천연두에 감염된 모든 사례 중 출혈성 천연두는 얼마의 비중을 차지하는가?   
...                                                 ...   
3745                      장수 개인의 특성 중 유일하게 다시 추가된 능력치는?   
3746                     우리나라 고인돌의 형식 중 지석묘는 무엇에 해당하는가?   
3748  중세 시기 삼경법이 이루어지는 밭 중 세균이나 몇몇 식물의 질소 고정 작용이 일어나...   
3917             차바하르항은 이란 항구 중 어느 국가의 제재를 유일하게 받지 않았나?   
3922                밤이 되면 공기가 위로 올라가는 곳은 육지와 바다 중 어디인가?   

                                            answers  document_id  
84    {'answer_start': [286], 'text': ['〈목포의 눈물〉']}        15686  
106         {'answer_start': [177], 'text': ['세균']}        57963  
144      {'answer_start': [365], 'text': ['에도 말기']}        17340  
180   {'answer_start': [165], 'text': ['이영희 막달레나']}        39168  
183     {'answer_start': [73], 'text': ['약 2% 정도']}         5536  
...                                             ...          ...  
3745        {'answer_start': [239], 'text': ['통솔']}        47415  
3746  {'answer_start': [241], 'text': ['바둑판식 고인돌']}        42506  
3748      {'answer_start': [85], 'text': ['세번째 밭']}        59627  
3917        {'answer_start': [180], 'text': ['미국']}        54152  
3922        {'answer_start': [627], 'text': ['바다']}        10486  

[109 rows x 5 columns]

In [21]:
is_multiple_choice_question(valid_df)

title                                            context  \
32      레닌주의  레닌주의에서 협동조합 개념을 이해할 때 주의해야 할 것은, 레닌이 주장하는 것은 자...   
35      착생식물  해양 착생식물의 생태는 지상 착생식물의 생태와는 다르다. 해양 생태계의 착생식물은 ...   
81       자수장  수를 놓을 때에 사용하는 기법을 문양의 형태에 따라 나누어 보면 선을 잇는 수, 면...   
82   코스모케라톱스  코스모케라톱스는 장식이 많은 두개골로 구분할 수 있는데, 어떤 공룡보다도 더 많은 ...   
121    최종생산물  한 나라의 경제활동의 일정기간 내의 모든 성과는 생산이라는 관점에서 본다면 여러 분...   
167   취업 빙하기  취업 빙하기가 진행되는 과정에서는 미취직자, 무직자가 급증하는데 이는 생산성 감소 ...   
236       티칼  현재 우리가 볼 수 있는 티칼의 모습은 펜실베이니아 대학교와 과테말라 정부의 협조 ...   

                                           question  \
32        레닌이 출간한 책 중 농민의 자발적 참여에 대한 내용이 포함되어있는 것은?   
35              해초나 조류 표면에서 자라는 유기체 중 가장 비율이 높은 것은?   
81           수를 놓을 때에 사용하는 기법 중 아주 작은 점으로 표현하는 방식은?   
82                    각룡류 중에서 가장 짧은 프릴을 가지고 있는 공룡은?   
121                    생산물 중 다른 물건을 만들기 위해 사용되는 것은?   
167  중고령자의 채취업 활성화로 경제 주체마저 옮겨가는 바람에 생기는 부작용 중 하나는?   
236                      공놀이 경기장 중 일부는 어디에 위치하고 있나?   

                                               answers  document_id  
32    {'answer_start': [312], 'text': ['『협동조합에 관하여』']}         5779  
35             {'answer_start': [180], 'text': ['조류']}        51188  
81             {'answer_start': [101], 'text': ['점수']}        19975  
82        {'answer_start': [250], 'text': ['코스모케라톱스']}        40380  
121        {'answer_start': [331], 'text': ['중간 생산물']}        59701  
167          {'answer_start': [365], 'text': ['소비감소']}        40838  
236  {'answer_start': [343], 'text': [''일곱 개의 신전 광장...        12190

## answer이 여러개인가?

In [22]:
def is_multiple_answers(df):
    answer_in_question = [
        row for _, row in df.iterrows() if len(row["answers"]["text"])!=1
    ]
    return pd.DataFrame(answer_in_question, columns=df.columns)

In [23]:
is_multiple_answers(train_df)

Empty DataFrame
Columns: [title, context, question, answers, document_id]
Index: []

In [24]:
is_multiple_answers(valid_df)

Empty DataFrame
Columns: [title, context, question, answers, document_id]
Index: []

## answer가 숫자만 있는 경우도 있는가?

In [10]:
def is_answer_digit(df):
    answer_digit = [
        row for _, row in df.iterrows() if row["answers"]["text"][0].isdigit()
    ]
    return pd.DataFrame(answer_digit, columns=df.columns)

In [11]:
is_answer_digit(train_df)

title  \
50                                     폴리스 박스   
425                                     중동신도시   
595                     마틸데 디 카노사 디 토스카나 여변경백   
646                                       여운형   
705                                    박근혜 정부   
776                                   겔라티 수도원   
854                          대한민국 아이스하키 국가대표팀   
939                          2002년 독일 연방의회 선거   
972                                 철원 갈말읍 토성   
1170                                  어번 네트워크   
1302                                  웨스트민스터궁   
1305                         1972년 독일 연방의회 선거   
1447               대한민국 대 독일 (2018년 FIFA 월드컵)   
1511                     밀양 표충사 삼층석탑 출토 유물 일괄   
1713                                      좌종당   
1839  동방귀형수 ~ Wily Beast and Weakest Creature   
2013                                     크노소스   
2559                                   3·1 운동   
2727                                구스타프 하이네만   
2730                                 히지카타 도시조   
2755                                 고바야시 세이조   
2861                                 상산군부인 김씨   
2869                               마흐무드 알카슈가리   
3060                            부다페스트 축제 관현악단   
3182                                   윌리엄 잭슨   
3259                                  쓰치야 히로시   
3379                                   마라의 죽음   
3538                               수요의 가격 탄력성   
3566                                  하템 트라벨시   
3581                                    갈로로마인   

                                                context  \
50    타디스\nBBC의 텔레비전 공상과학 드라마 시리즈 《닥터 후》에서는 타임머신인 타디...   
425   중동신도시(中洞新都市)는 경기도 부천시 중동, 상동 일대에 조성된 신도시이다. 중동...   
595   마틸데의 아버지는 레지오, 모데나, 만토바, 브레시아, 페라라 등에 영지를 가진 이...   
646   1907년부터 1910년까지 초기 민중교회였던 승동교회 (서울특별시 종로구 인사동 ...   
705   2013년 대학수학능력시험 세계지리 과목의 문제에 오류가 있다고 일부 학계와 수험생...   
776   겔라티 수도원은 조지아 서부의 이메레티의 쿠타이시 근교에 있는 수도원 단지이다. 그...   
854   한국에 아이스하키가 도입된 것은 1920년대로 알려져 있으나, 세계 아이스하키 선수...   
939   2002년 독일 총선거는 전체 603석의 독일 연방의회 의원들을 선출하기 위해 20...   
972   토성리 마을의 밭을 둘러 싸고 있는 흙으로 쌓은 성과 군탄리 마을의 옛 풍전역터에 ...   
1170  1987년에 일본국유철도의 분할 민영화가 진행되고, 서일본 여객철도는 오사카 권을 ...   
1302  웨스트민스터 홀은 현재 궁전 내부에 있는 가장 오래된 방이다. 윌리엄 2세에 의해 ...   
1305  1972년 서독 총선거는 전체 496석의 서독 연방의회의 의원들과 서베를린에 할당된...   
1447  대한민국과 독일이 FIFA 월드컵에서 3번째로 맞붙은 경기이자 양 팀의 4번째 A매...   
1511  밀양 표충사 삼층석탑 출토 유물 일괄은 1995년 실시된 삼층석탑의 해체보수 과정에...   
1713  염군의 난\n1866년(동치 5년)에는 섬감총독으로 전임하여, 회족(이슬람교도)인 ...   
1839  적을 쓰러뜨리면 떨어뜨리는 '동물령 아이템'을 5개 모음으로써 발동된다. 동물령 아...   
2013  크노소스는 신석기 시대(기원전 7000년경)부터 사람이 거주해왔다. 크레타에는 동굴...   
2559  3·1독립선언서\n 1919년 3월 1일 오후 2시 에 모이기로 했던 조선의 민족대...   
2727  학생으로서 하이네만은 바이마르 공화국을 강하게 후원한 자유적 독일 민주당의 학생 기...   
2730  10월 20일, 에조치 와시노키에 상륙 후 도시조는 문도군 총독이 되어서 고료카쿠(...   
2755  1920년에는 다시 영국의 주재무관으로 파견되었고, 이때 영국의 항공전문가 윌리엄 ...   
2861  처음에는 시신을 그해 8월 8일 경기도 금천 하북면 번대방리(현, 서울특별시 동작구...   
2869  마흐무드 알카슈가리는, 그의 니스바 알카슈가리(الكاشغري)에서 알 수 있듯, ...   
3060  부다페스트 축제 관현악단(Budapesti Fesztiválzenekar, Buda...   
3182  잭슨은 영국의 컴벌랜드 카운티에서 태어나 부모가 죽은 후에, 사우스캐롤라이나의 찰스...   
3259  우스다 고등학교 시절에는 2학년 때부터 에이스로서 활약하여 1961년 추계 호쿠신에...   
3379  ‘마라의 죽음’은 널리 찬양을 받았고 공포 정치 시대의 지도자들은 오리지널 작품의 ...   
3538  수요의 가격탄력성은 각 요인들의 변화가 수요량이 얼마나 민감하게 반응하게 되는지를 ...   
3566  트라벨시는 스팍스에서 자랐으며 그의 축구 인생도 그곳에 자리잡은 CS 스팍시앙에서 ...   
3581  로마는 갈리아를 3개 속주로 나누어 지배했고, 행정구역으로는 과거 갈리아 부족들의 ...   

                                           question            id  \
50         BBC의 파란 경찰 박스 디자인 상표권에 대한 법적 논란이 종료된 해는?  mrc-0-002942   
425                  중동신도시 개발에 기관 몇 곳이 적극적으로 뛰어들었나?  mrc-0-002948   
595                            마틸데의 첫째 언니가 사망한 년도는?  mrc-0-005249   
646                           여운형의 노비들이 자유를 찾았던 해는?  mrc-0-002644   
705                         수능에서 가장 많은 오류가 발생한 연도는?  mrc-0-002421   
776                     갈라티 수도원이 세계 문화 유산으로 선정된 해는?  mrc-1-000947   
854             대한민국이 5개국을 꺾고 디비전 1A로 승급될 수 있었던 해는?  mrc-1-000253   
939   2002년 9월 22일에 시행된 독일 총선거에서 뽑고자 했던 연방의회 의원 수는?  mrc-0-000782   
972                        이호민이 과거시험에 합격한 것은 몇년도인가?  mrc-0-002562   


In [12]:
is_answer_digit(valid_df)

Empty DataFrame
Columns: [title, context, question, id, answers, document_id, __index_level_0__]
Index: []

## question에 물음표가 항상 존재하는가?

In [25]:
def is_not_q_marks(df):
    not_q_marks = [
        row for _, row in df.iterrows() if "?" not in row["question"]
    ]
    return pd.DataFrame(not_q_marks, columns=df.columns)

In [26]:
is_not_q_marks(train_df)

title                                            context  \
295  피아노 협주곡 5번 (베토벤)  자신의 후견자 겸 제자인 루돌프 대공을 위해, 1809년에 프랑스 군대의 포격이 쏟...   

                     question                                         answers  \
295  피아니스트이자 출판업을 하는 베토벤의 친구는  {'answer_start': [432], 'text': ['요한 B. 크라머']}   

     document_id  
295         7870

In [27]:
is_not_q_marks(valid_df)

Empty DataFrame
Columns: [title, context, question, answers, document_id]
Index: []

## 답변 가능성
- 답변 가능한지 아닌지에 대한 라벨을 추가한다면 No Answer를 더 잘 학습할 수 있을것.

In [28]:
def is_answer_blank(df):
    answer_in_question = [
        row for _, row in df.iterrows() if row["answers"]["text"]==""
    ]
    return pd.DataFrame(answer_in_question, columns=df.columns)

In [29]:
is_answer_blank(train_df)

Empty DataFrame
Columns: [title, context, question, answers, document_id]
Index: []

In [30]:
is_answer_blank(valid_df)

Empty DataFrame
Columns: [title, context, question, answers, document_id]
Index: []

## Rule-base 에러 탐지

In [14]:
t = train_df[["document_id", "question", "answers"]].copy()
t["answers"] = t["answers"].map(lambda x: x["text"][0])

v = valid_df[["document_id", "question", "answers"]].copy()
v["answers"] = v["answers"].map(lambda x: x["text"][0])

In [23]:
from konlpy.tag import Kkma
import re
import pandas as pd

# Kkma 형태소 분석기 사용
kkma = Kkma()


def check_answer(df):
    wrong_data = []
    keywords = ["언제", "날짜", "누구", "어디", "무엇", "왜", "어떻게"]

    # DataFrame에서 각 행에 대해 검사
    for index, row in df.iterrows():
        question = row["question"]
        answer = row["answers"]

        # 질문에서 키워드 추출
        keyword = None
        tokens = kkma.pos(question)
        for word, tag in tokens:
            if word in keywords:
                keyword = word
                break

        if keyword is None:
            continue  # 질문에서 키워드가 발견되지 않으면 넘어감

        # 답변 유형 판단
        if keyword in ["언제", "날짜"]:
            # 숫자 있는지 체크
            if not bool(re.search(r"\d", answer)):
                wrong_data.append(row)
        elif keyword == "누구":
            # 사람 이름인지 판단 (KoNLPy를 사용하여 고유명사 추출)
            answer_tokens = kkma.pos(answer)
            person_found = False
            for word, tag in answer_tokens:
                if tag == "NNP" or tag == "NNG":  # 고유명사 또는 일반명사
                    person_found = True
                    break

            if not person_found:
                wrong_data.append(row)

    return pd.DataFrame(wrong_data, columns=df.columns)

def check_date_answer(df):
    def contains_keyword(text, keywords):
        for keyword in keywords:
            if keyword in text:
                return True
        return False

    wrong_data = []
    date_keywords = ["언제","날짜","해","년","월","일","때","시기","연도","기간"]

    # DataFrame에서 각 행에 대해 검사
    for index, row in df.iterrows():
        question = row["question"]
        answer = row["answers"]
        if contains_keyword(question,date_keywords):
            continue
        if answer.isdigit() or contains_keyword(answer,date_keywords):
            wrong_data.append(row)
    return pd.DataFrame(wrong_data, columns=df.columns)

In [55]:
wrong_train = check_answer(t)
wrong_train

document_id                                      question       answers
144         17340                일본의 원정 중 가장 최근의 원정은 언제 실행되었는가?         에도 말기
167         39694                         푸르바더는 언제 세상 밖으로 나왔는가?     에흐네가 죽은 뒤
200         39711                                만덕사는 언제 설립되었나?         고려 전기
217         33762    중국에서 쿨리가 공식적으로 해외에 나갈 수 있게 되었던 시점은 언제부터인가?  베이징 조약 체결 이후
316         20216                                  호남선은 언제 생겼나?        일제 강점기
339         18041                      임영관 건물이 학교로 사용된 것은 언제인가?          일제시대
369         10698            세카이와 코토노하가 마코토에게 줄 선물은 언제를 위한 것인가?         크리스마스
426         23287                           탑에 시멘트를 바른 것은 언제인가?         일제강점기
459         35475                          부여석조가 만들어진 시기는 언제인가?          백제시대
549         29601              사이판의 전선과 하수구 파이프는 언제 수리하기 시작했는가?  태풍 초이완이 지난 후
660         46378                   네이든이 몰래 좋아하고 있었던 인물은 누구인가요?             닉
716         57423           기갑 세력을 이용하여 호르티에게 항복을 받아낸 사람은 누구인가?         슈코르체니
781         36795         튜더 왕가의 정통성에 해를 가했던 사람들은 언제 죽음을 당하였나요?      튜더 왕조 시대
783         26355                 조지아인과 오세트인이 화합을 이루웠던 때는 언제인가?       소비에트 시대
799         15951                          남상철의 아버지는 언제 살아남았나요?          병인박해
801         15361                  빔비사라의 아들이며 앙가를 다스린 사람은 누구인가?        아자타샤트루
950         16451                      저고리의 길이가 짧아진 시기는 언제부터인가?          고려 말
977         28351                            순무 제도가 생긴 것은 언제인가?        영락제 시기
1079        18910                      밀본을 얕잡아 본 인물은 누구의 친구였나요?            수천
1155         8950                  반 아이들에게 세리가 누구를 좋아한다고 들통났나요?             욱
1159        39874                   프라이드 하우스는 누구에 의해 처음 설치되었는가?          딘 넬슨
1411        51793                     유세군은 누구와의 관계 때문에 시를 지었나요?            곤막
1735        29587               조금씩 유럽사회가 계급사회로 접어들게 된 때는 언제인가?         중세 시대
1950        36795                       레지널드 폴의 어머니는 언제 사망하였나요?      튜더 왕조 시대
1983        28300              언론, 교육 등에 힘썼던 가톨릭 조직은 언제 활동하였는가?          일제시대
2562        43065  성매매 종사자가 계약에 묶일 필요없이 언제든 그만 둘 수 있음을 의미하는 말은?        자율적 결정
2586        20027          도조 히데키는 누구의 죽음 이후에 통제파의 주요 인물이 되었나요?           나가타
2695        25919                   공자 돌이 임금의 자리에 올랐던 시기는 언제인가?           기해일
2897        20170                       CIS 탄생의 계기가 된 인물은 누구인가?            옐친
2978        30159                         철불상이 많이 제작된 시대는 언제인가?       통일신라 후기
3016        26472                      후두의 영어표현은 누구에게서 비롯된 것인가?            아담
3115        26445                       통일된 조지아 왕국의 초대 왕은 누구인가?          바그라트
3141        13020                             이중창은 언제를 대비한 것인가?           겨울철
3235        45731                  아르타크세르크세스를 독으로 살해한 인물은 누구인가?          바고아스
3256        20027         도조 히데키는 누구의 사망 이후에 비황도파의 중심 인물이 되었나요?           나가타
3403        41390                      아버지를 존중하지 않았던 인물은 누구인가요?             킴
3417        22186              임산부가 태몽을 꾸는 시기는 주로 임신전후와 또 언제인가?          분만직전
3426         5866                   번오기의 진군에 의해 살해 당한 대장은 누구인가?            호첩
3618        45853          야야티의 아들이자 파우라바 왕국을 건국한 민족의 선조는 누구인가?            푸루
3648        28469        타티키오스가 비밀협상에 나선 것은 누구의 기록을 통해 알 수 있나요?        안나 콤네나
3827        47270                               지석묘는 언제 만들어졌나요?         청동기시대

In [24]:
wrong_valid = check_answer(v)
wrong_valid

document_id                                         question   answers
57         45428                              당산제를 시행하던 날은 언제였는가?  매년 음력 정월
78         28290  대한민국 민법에서 채권의 권리가 누구에게 있으면 채권이 사라지지 않는다고 명시하는가?      제 3자
104        10728                                콘스탄스를 죽인 것은 누구인가?      350년
175        41382                  나오치카는 주살의 위험을 피해 누구의 영지로 도망갔나요?       다케다
224         9420             배현경과 신숭겸 등이 일을 꾀하여 찾아간 인물의 부인은 누구인가?      제1부인

In [25]:
wrong_train = check_date_answer(t)
wrong_train

document_id                                 question  \
15          47810                권번에서 요릿집에 제공한 기생 명부의 이름은?   
58          23216                       많은 갈등을 불러온 툴민의 책은?   
85           7119                     에케르트가 시행착오를 겪었던 국가는?   
99          44197                          소프트 머신의 이전 명칭은?   
128         59489                  시네마토그래프 유료 상영회가 개최된 날은?   
...           ...                                      ...   
3713        19657            보로메오 가문의 문장은 어떤 의미를 지니고 있었나요?   
3814        34647  학교형태의 평생교육시설이 어떠한 목적으로 이용되는 것을 막아야 하는가?   
3824        10511              프로비던스와 함께 가장 먼저 노예제를 없앤 곳은?   
3845         5016                 강희제는 장병 간의 관계가 어떻다고 믿었나?   
3890        51962             대교파출소는 어떤 사람한테서 장난 전화를 받았는가?   

                               answers  
15                               '초일기'  
58    ≪인간의 이해: 개념의 집단적 사용 및 진화 (1972)≫  
85                                  일본  
99                            와일드 플라워즈  
128                      1895년 12월 28일  
...                                ...  
3713                              삼위일체  
3814            고등학교를 보다 일찍 졸업하기 위한 수단  
3824                          로드아일랜드 주  
3845                              일심동체  
3890                     김현정 양을 살해한 범인  

[157 rows x 3 columns]

In [26]:
wrong_valid = check_date_answer(v)
wrong_valid

document_id                             question  \
11          8777         인형사'를 만들어낸 것으로 추측되는 사업의 이름은?   
17         57479           이흥구의 사법시험 이야기를 기사로 작성한 곳은?   
51         18819                 슈트레제만이 이끌었던 당은 무엇인가?   
60          8823            벨기에가 리에주 전투에서 맞서 싸운 상대방은?   
77         11517               할더가 히틀러의 공격안을 반대한 결과는?   
101        46019             메테스키가 최초로 폭탄을 설치하였던 지역은?   
104        10728                    콘스탄스를 죽인 것은 누구인가?   
109        24978  어느정도 규모 이상의 대회에서 깃털 셔틀콕만을 사용하는 이유는?   
197         6690                 아카키우스 분열은 얼마나 유지됐는가?   
209        43204                    마오를 직접적으로 죽인 사람은?   

                                               answers  
11                                   '일급 비밀 프로젝트 2501'  
17                                              〈중앙일보〉  
51                                               독일인민당  
60                                                 독일군  
77                                                  독일  
101                                                맨해튼  
104                                               350년  
109  깃털 셔틀콕의 타구감을 선호하고, 또한 플라스틱보다 깃털 셔틀콕이 정교한 컨트롤을 ...  
197                                                35년  
209                                            일등병 토머스

## 정답 문장 추출

In [113]:
# 정답이 포함된 문장을 추출하는 함수
def extract_answer_sentence(row):
    context = row["context"]
    answer_text = row["answers"]["text"][0]
    answer_start = row["answers"]["answer_start"][0]

    # 정답의 끝 위치 계산
    answer_end = answer_start + len(answer_text)

    # 문장의 시작 부분 찾기 (answer_start부터 왼쪽으로)
    sentence_start = answer_start
    while sentence_start > 0 and context[sentence_start - 1] not in [
        ".",
        "?",
        "!",
        "\n",
    ]:  # 문장 부호 및 개행 확인
        sentence_start -= 1

    # 문장의 끝 부분 찾기 (answer_end부터 오른쪽으로)
    sentence_end = answer_end
    while sentence_end < len(context) and context[sentence_end] not in [
        ".",
        "?",
        "!",
        "\n",
    ]:  # 문장 부호 및 개행 확인
        sentence_end += 1

    # 완전한 문장을 추출
    extracted_sentence = context[sentence_start:sentence_end].strip()

    return extracted_sentence

In [117]:
valid_df["answer_sentence"] = valid_df.apply(extract_answer_sentence, axis=1)

In [118]:
train_df["answer_sentence"] = train_df.apply(extract_answer_sentence, axis=1)

## 눈으로 직접 확인하기: QA 쌍 보기

In [122]:
pd.options.display.max_colwidth = 2000
pd.set_option("display.max_rows", 20)

In [119]:
v = valid_df[["document_id", "answer_sentence", "question", "answers"]].copy()
v["answers"] = v["answers"].map(lambda x: x["text"][0])
v.to_json("./valid.json", force_ascii=False, orient = 'records', indent=4)

In [127]:
'''
10728: 360년 -> 마그넨티우스
9420: 제1부인 -> 신혜왕후 또는 유씨
'''

'\n10728: 360년 -> 마그넨티우스\n9420: 제1부인 -> 신혜왕후 또는 유씨\n'

In [120]:
t = train_df[["document_id", "answer_sentence", "question", "answers"]].copy()
t["answers"] = t["answers"].map(lambda x: x["text"][0])
t.to_json("./train.json", force_ascii=False, orient="records", indent=4)

In [123]:
valid_df[valid_df["document_id"] == 40838]

title  \
167  취업 빙하기   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       context  \
167  취업 빙하기가 진행되는 과정에서는 미취직자, 무직자가 급증하는데 이는 생산성 감소 및 경기 침체 가속화, 생산성 감소, 소비 감소, 저출산 고령화, 만혼화, 미혼화 등의 원인이 된다. 이들 미취직자, 무직자들 중에는 구직 의욕 자체가 없는 니트족과 구직 의욕은 있으나 쉽게 일자리를 구하지 못하는 실업자 또는 실직자로 분류된다. 구직을 포기한 미취직자는 잠재적 사회불안 요인 및 반사회적 범죄의 우려가 있으며 모든 미취직자, 무직자를 사회부적응으로 규정할 수는 없으나, 이들 중에는 은둔형 사회부적응자가 되기도 한다. 신규자 고용 감소와 경력자 선호 등으로 중고령자의 재취업이 활성화되면서 경제 주체가 청년층에서 중고령자, 중장년층으로 옮겨갔으며 이는 소비감소 등의 다른 부작용도 유발되고 있다.\n\n업무를 이해하지 못한 상태에서 취직했거나, 학교에서 가르친 현장 실무와는 동떨어진 이론 위주의 교육 및 경제적, 정서적 문제 등으로 취직 후 1년 안에 그만두거나 이직하는 사례가 증가한다. 또한 생산 현장에 맞지 않는 지나친 고학력 인력의 과잉 공급으로, 전공 분야로 취업했으나 현장에서 오래 근무한 이들과의 감정적 대립으로 쉽게 자퇴하거나 이직하는 일도 종종 발생한다. 그밖에 학교에서의 교육이 지나친 암기 강요 또는 암기 위주의 이론 교육에 치중했던 점 또한 학생과 신규 졸업자의 현장 진출을 어렵게 만드는 구조적 문제점의 하나로 지적된다.\n\n취업 빙하기의 문제점으로 청년 구직자의 눈높이 문제가 지적된다. 그러나 학교에서 산업 현장의 수요에 맞는 인력을 양성하지 못하는 것과, 현장 수요에 맞지 않는 지나친 고학력화 및 대학교, 전문대, 대학원 등 고등 학력 기관의 남발로 학생 및 신규 졸업자의 일자리에 대한 기대도는 높아진 반면, 기업 현장에서는 즉시 투입될 수 있는 인물만을 선호하여 구직자와 구인자 간의 갈등의 양상도 띄고 있다. 즉시 현장 실무에 투입될 수 있는 인물을 가리켜 즉전인재라 한다. 한편 산업 현장에서는 신규자 실무 교육 시간이 없다는 이유로 방치하거나, 과도한 업무를 맡기거나 심리적 스트레스를 유발케 하여 의도적으로 그만두도록 만드는 유도퇴직 등의 현상도 등장하였다.   

                                           question  \
167  중고령자의 채취업 활성화로 경제 주체마저 옮겨가는 바람에 생기는 부작용 중 하나는?   

                                       answers  document_id  \
167  {'answer_start': [365], 'text': ['소비감소']}        40838   

                                                                                        answer_sentence  
167  신규자 고용 감소와 경력자 선호 등으로 중고령자의 재취업이 활성화되면서 경제 주체가 청년층에서 중고령자, 중장년층으로 옮겨갔으며 이는 소비감소 등의 다른 부작용도 유발되고 있다

In [129]:
train_df[train_df["document_id"] == 28730]

title  \
145   첩혈쌍웅   
1173  첩혈쌍웅   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        context  \
145   살인 청부업자인 아장(주윤발)은 살인청부를 실행하다 무고한 여가수 제니(엽청문)의 눈을 실명하게 만든다. 이에 대한 자책감으로 제니의 주변을 맴돌던 아장은 제니와 사랑에 빠진다. 한편 과감한 결단력으로 수사를 진행하는 이응(이수현)은 경찰 상부로부터 이때문에 늘 문책을 당하는 입장이다. 한편 주윤발은 제니의 각막 수술 비용을 벌기 위해 토니왕이라는 재력가이자 암흑가와도 연관성이 있는 자의 살인 청부를 맡고, 이응은 상부의 명령으로 토니왕의 경호를 맡지만 그다지 내켜하지는 않는다. 아장이 토니왕을 저격하고 도주하자 이응은 이를 추격하는데 침착함을 잃지 않으면서 무고한 희생을 피하는 아장에게 이응은 강한 인상을 받는다. 한편 아장은 토니왕의 암살을 지시한 토니왕의 조카 헤이왕으로부터 돈을 받기는커녕 오히려 살해 위협을 당해 도주하게된다. 이응은 제니에게 접근해서 아장을 체포하러 시도하지만 실패하고 문책당한다. 아장을 노리고 잠복해있던 이응의 동료경관과 헤이왕의 부하들은 충돌해서 이응의 동료경관은 사망하게되고 이에 분노한 이응은 아장을 추격하지만 아장의 의리, 우정, 사랑에 감복해 아장의 편이 된다. 한편 아장의 살인청부 대리인 역할을 하던 아장의 친구가 아장이 받아야 할 돈을 받아내려다 헤이왕의 부하들에게 살해당하자 아장은 복수를 다짐하고 직접 자신을 죽이러 부하들과 찾아온 헤이왕과 총격전을 벌인다. 그 과정에서 헤이왕의 부하들은 모두 죽고 제니는 헤이왕에게 인질로 잡힌다. 제니를 구하려다 아장은 온몸에 총격을 맞고 결국 사망하고, 헤이왕은 도주하던 끝에 경찰들을 보자 목숨을 건지기 위해 자수한다. 그러나 이를 쫓아온 이응은 경찰들에게 체포된 헤이왕에게 총을 쏘면서 아장을 그리워한다.   
1173  살인 청부업자인 아장(주윤발)은 살인청부를 실행하다 무고한 여가수 제니(엽청문)의 눈을 실명하게 만든다. 이에 대한 자책감으로 제니의 주변을 맴돌던 아장은 제니와 사랑에 빠진다. 한편 과감한 결단력으로 수사를 진행하는 이응(이수현)은 경찰 상부로부터 이때문에 늘 문책을 당하는 입장이다. 한편 주윤발은 제니의 각막 수술 비용을 벌기 위해 토니왕이라는 재력가이자 암흑가와도 연관성이 있는 자의 살인 청부를 맡고, 이응은 상부의 명령으로 토니왕의 경호를 맡지만 그다지 내켜하지는 않는다. 아장이 토니왕을 저격하고 도주하자 이응은 이를 추격하는데 침착함을 잃지 않으면서 무고한 희생을 피하는 아장에게 이응은 강한 인상을 받는다. 한편 아장은 토니왕의 암살을 지시한 토니왕의 조카 헤이왕으로부터 돈을 받기는커녕 오히려 살해 위협을 당해 도주하게된다. 이응은 제니에게 접근해서 아장을 체포하러 시도하지만 실패하고 문책당한다. 아장을 노리고 잠복해있던 이응의 동료경관과 헤이왕의 부하들은 충돌해서 이응의 동료경관은 사망하게되고 이에 분노한 이응은 아장을 추격하지만 아장의 의리, 우정, 사랑에 감복해 아장의 편이 된다. 한편 아장의 살인청부 대리인 역할을 하던 아장의 친구가 아장이 받아야 할 돈을 받아내려다 헤이왕의 부하들에게 살해당하자 아장은 복수를 다짐하고 직접 자신을 죽이러 부하들과 찾아온 헤이왕과 총격전을 벌인다. 그 과정에서 헤이왕의 부하들은 모두 죽고 제니는 헤이왕에게 인질로 잡힌다. 제니를 구하려다 아장은 온몸에 총격을 맞고 결국 사망하고, 헤이왕은 도주하던 끝에 경찰들을 보자 목숨을 건지기 위해 자수한다. 그러나 이를 쫓아온 이응은 경찰들에게 체포된 헤이왕에게 총을 쏘면서 아장을 그리워한다.   

                               question  \
145                  토니왕의 조카에게 총을 쏜 자는?   
1173  제니가 아장에 의해 장애를 가지게 된 신체 부위는 어디인가?   

                                      answers  document_id  \
145   {'answer_start': [804], 'text': ['이응']}        28730   
1173    {'answer_start': [46], 'text': ['눈']}        28730   

                                                answer_sentence  
145             그러나 이를 쫓아온 이응은 경찰들에게 체포된 헤이왕에게 총을 쏘면서 아장을 그리워한다  
1173  살인 청부업자인 아장(주윤발)은 살인청부를 실행하다 무고한 여가수 제니(엽청문)의 눈을 실명하게 만든다

### 결론 (train 310개, val 240개 약 550개의 데이터 직접 체크함)
1. 생각보다 데이터 품질이 좋음. 전반적으로 큰 오류를 찾기 어려웠다.
2. 외부 데이터를 사용하는 것 만큼 대회 제공 데이터를 증강하는 것이 유효할 것이라고 생각이 든다.

### 추가적으로 검증해보면 좋을 내용
1. question의 난이도 평가 (question 유형, question에 사용된 어휘가 context와 얼마나 일치하는지 등으로)